aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,359 @@
+            per_item_write: 100,
+            per_byte_read: 1,
+            per_byte_create: 1,
+            per_byte_write: 1,
+        });
+    }
+
+    fun validate_usage_config(config: &UsageGasConfig) {
+        assert!(config.target_usage > 0, error::invalid_argument(EZERO_TARGET_USAGE));
+        assert!(config.target_usage <= 18446744073709551615 / BASIS_POINT_DENOMINATION, error::invalid_argument(ETARGET_USAGE_TOO_BIG));
+        validate_curve(&config.read_curve);
+        validate_curve(&config.create_curve);
+        validate_curve(&config.write_curve);
+    }
+
+    fun validate_curve(curve: &GasCurve) {
+        assert!(curve.max_gas >= curve.min_gas, error::invalid_argument(EINVALID_GAS_RANGE));
+        let points = &curve.points;
+        let len = vector::length(points);
+        let i = 0;
+        while (i <= len) {
+            let cur = if (i == 0) { &Point {x: 0, y: 0} } else { vector::borrow(points, i - 1) };
+            let next = if (i == len) { &Point {x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION} } else { vector::borrow(points, i) };
+            assert!(cur.x < next.x && cur.y <= next.y, error::invalid_argument(EINVALID_MONOTONICALLY_NON_DECREASING_CURVE));
+            i = i + 1;
+        }
+    }
+
+    fun calculate_gas(max_usage: u64, current_usage: u64, curve: &GasCurve): u64 {
+        let capped_current_usage = if (current_usage > max_usage) max_usage else current_usage;
+        let points = &curve.points;
+        let num_points = vector::length(points);
+        let current_usage_bps = capped_current_usage * BASIS_POINT_DENOMINATION / max_usage;
+
+        // Check the corner case that current_usage_bps drops before the first point.
+        let (left, right) = if (num_points == 0) {
+            (&Point {x: 0, y: 0}, &Point {x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION})
+        } else {
+            let (i, j) = (0, num_points - 1);
+            while (i < j) {
+                let mid = j - (j - i) / 2;
+                if (current_usage_bps < vector::borrow(points, mid).x) {
+                    j = mid - 1;
+                } else {
+                    i = mid;
+                };
+            };
+            if (current_usage_bps < vector::borrow(points, 0).x) {
+                (&Point {x: 0, y: 0}, vector::borrow(points, 0))
+            } else {
+                // Check the corner case that current_usage_bps drops after the last point.
+                (
+                    vector::borrow(points, i),
+                    if (i == num_points - 1) {
+                        &Point { x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION }
+                    } else {
+                        vector::borrow(points, i + 1)
+                    }
+                )
+            }
+        };
+
+        curve.min_gas + (curve.max_gas - curve.min_gas) * (left.y + (current_usage_bps - left.x) * (right.y - left.y) / (right.x - left.x)) / BASIS_POINT_DENOMINATION
+    }
```

Comment:
@davidiw @vgao1996 Can our SMART smart contract engineers help review the code to reassure us that it will not abort in any cases?
1. line 209 would not overflow because, line 186 checks `max_usage` * 10000 <= u64::MAX
2. Please review the binary search logic is valid.
3. Please check 239 will not overflow or underflow( such as current_usage_bps - left.x).
Some background: 
in validate_curve(), we check 0 <= point.x, point.y <= BASIS_POINT_DENOMINATION, x is strictly monotonically increasing while y is monotonically non-decreasing;
max_gas * BASIS_POINT_DENOMINATION <= u64::MAX.

Reply:
Yeah when I read this part of the code previously I also checked that the code would not overflow/underflow